platform/cert/cert_manager.go

```diff
@@ -98,11 +106,54 @@ func (c *certManager) UpdateCertificates(certs string) error {
 	}
 	c.logger.Debug(c.logTag, "Wrote %d new certificate files", len(slicedCerts))
 
-	_, _, _, err = c.runner.RunCommand(c.updateCmdPath, c.updateCmdArgs...)
-	if err != nil {
-		return bosherr.WrapError(err, "Updating certificates")
+	// For Ubuntu OS, update-ca-certificates occasionally hangs, which results
+	// in bosh-agent failure. A retry normally solves this issue. We kill the process
+	// if it runs over given time limit and retry for 3 times until we throw error.
+	if c.updateTimeout > 0 {
+		command := boshsys.Command{
+			Name: c.updateCmdPath,
+			Args: c.updateCmdArgs,
+		}
+
+		for i := 1; i < 4; i++ {
+			c.logger.Debug(c.logTag, "Try to update new certificate files with retry, take %d", i)
+
+			process, err := c.runner.RunComplexCommandAsync(command)
+			if err != nil {
+				return bosherr.WrapError(err, "Updating certificates")
+			}
+
+			resultChannel := process.Wait()
+
+			select {
+			case <-time.After(time.Duration(c.updateTimeout) * time.Second):
+				err = process.TerminateNicely(5 * time.Second)
```

Comment:
do we have a test in exec_cmd_runner that makes sure that TerminateNicely does not error out when it's trying to kill process that's already gone?